### DISCUSSION TITLE
layer(): passing through modifier keys

### DISCUSSION THREAD

**Discussion by sohocoke** [evan-liu/karabiner.ts#59] [Q&A] [answered]
Amazing project! 

I created a layer with arrow keys that I activate with F, so I have: 
```
    layer('f', 'f-arrow-mode').manipulators([
      map('i').to('↑'),
      map('j').to('←'),
      map('l').to('→'),
      map('k').to('↓'),
    ])
```

Is there a way to let these mappings work orthogonally with modifiers such as Shift or Option? Would I even be able to remap such modifiers to D, S etc? Would be awesome to be able to hold down S+D+F then hit J to select words backwards just with the homerow keys.

**Comment by evan-liu** [comment]
You can try (optional 'any' modifiers):

```typescript
  layer('f', 'f-arrow-mode').manipulators([
    map('i', undefined, 'any').to('↑'),
    map('j', undefined, 'any').to('←'),
    map('l', undefined, 'any').to('→'),
    map('k', undefined, 'any').to('↓'),
  ]),
```

or 

```typescript
  layer('f', 'f-arrow-mode').manipulators([
    withMapper<FromKeyParam, ToKeyParam>({
      i: '↑',
      j: '←',
      l: '→',
      k: '↓',
    })((k, v) => map(k, undefined, 'any').to(v)),
  ]),
```

**Comment by evan-liu** [comment]
For S an D key you can try 

```typescript 
  rule('Letter to Modifiers').manipulators([
    map('d').to('left_shift').toIfAlone('d'),
    map('s').to('left_option').toIfAlone('s'),
  ]),
```

**Comment by sohocoke** [comment]
Thanks for all these tips! I found one unexpected issue with the idea of using F as a layer however – I discovered that when I type, I would accidentally trigger the new layer for words such as 'find'. I think it's because the F key up doesn't occur before the I key down when I type fast. 

I discovered https://ke-complex-modifications.pqrs.org/json/mouse_keys_mode_v4.json which is conceptually identical to a configuration I'm after, and a cursory read through makes me think that if the layer() method could generate from.simultaneous and simultaneous_options for all the layer key + mapped key combinations, it could resolve the issue.

I'll try to test it out when I have some spare time.

**Reply by evan-liu** [reply, marked as answer]
I think you are looking for `simlayer()`

**Reply by sohocoke** [reply]
Sweet, thanks a lot! 'It just works' 👍

**Reply by sohocoke** [reply]
Would there be a way to setup a simlayer to allow modifiers? 

Once I activate the simlayer by e.g. holding F then hitting J, I can add Shift to select left. But holding F + Shift outputs F before I have a chance to hit J and trigger the simlayer. Same with F + Cmd, then J – as soon as I press Cmd the config seems to determine I don't want the simlayer.

**Reply by evan-liu** [reply]
I just [made 2 changes](https://github.com/evan-liu/karabiner.ts/releases/tag/1.11.0) to `simlayer()`: 

1. Add `{ optional: ["any"] }` modifiers by default. 
2. Add `options()` method to override [default `simultaneous_options`](https://github.com/evan-liu/karabiner.ts/blob/main/src/config/simlayer.ts#L42-L45). 

```typescript
{
    detect_key_down_uninterruptedly: true,
    key_down_order: 'strict',
    key_up_order: 'strict_inverse',
    key_up_when: 'any',
}
```

You can try different options to see if what would work for you like:

 ```typescript
simlayer().options({ detect_key_down_uninterruptedly: false })
```

You can find [the details of the options here](https://karabiner-elements.pqrs.org/docs/json/complex-modifications-manipulator-definition/from/simultaneous-options/). 

Please update (`npm run update`) and let me know how it goes. Thanks for your time trying out the project.

**Reply by sohocoke** [reply]
Assuming I am to run `npm run update` in the `karabiner.ts.examples` project (since there is no update script in the `karabiner.ts` project): I'm finding `node_modules/karabiner.ts` is not updating from 1.4.1. 

Terminal output:
```
user@host ~/d/s/p/karabiner.ts.examples> npm run update --verbose                                                                                                      main!?
npm info it worked if it ends with ok
npm verb cli [
npm verb cli   '/Users/user/.asdf/installs/nodejs/14.21.2/bin/node',
npm verb cli   '/Users/user/.asdf/installs/nodejs/14.21.2/bin/npm',
npm verb cli   'run',
npm verb cli   'update',
npm verb cli   '--verbose'
npm verb cli ]
npm info using npm@6.14.17
npm info using node@v14.21.2
npm verb run-script [ 'preupdate', 'update', 'postupdate' ]
npm info lifecycle karabiner.ts.examples@~preupdate: karabiner.ts.examples@
npm info lifecycle karabiner.ts.examples@~update: karabiner.ts.examples@

> karabiner.ts.examples@ update /Users/user/dev/src/public/karabiner.ts.examples
> npm update karabiner.ts

npm info it worked if it ends with ok
npm verb cli [
npm verb cli   '/Users/user/.asdf/installs/nodejs/14.21.2/bin/node',
npm verb cli   '/Users/user/.asdf/installs/nodejs/14.21.2/bin/npm',
npm verb cli   'update',
npm verb cli   'karabiner.ts'
npm verb cli ]
npm info using npm@6.14.17
npm info using node@v14.21.2
npm verb npm-session d602b75849887115
npm verb update computing outdated modules to update
npm verb exit [ 0, true ]
npm timing npm Completed in 246ms
npm info ok 
npm verb lifecycle karabiner.ts.examples@~update: unsafe-perm in lifecycle true
npm verb lifecycle karabiner.ts.examples@~update: PATH: /Users/user/.asdf/installs/nodejs/14.21.2/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/Users/user/dev/src/public/karabiner.ts.examples/node_modules/.bin:/Users/user/.asdf/plugins/nodejs/shims:/Users/user/.asdf/installs/nodejs/14.21.2/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Applications/VMware Fusion.app/Contents/Public:/Applications/Little Snitch.app/Contents/Components:/opt/X11/bin:/Applications/Server.app/Contents/ServerRoot/usr/bin:/Applications/Server.app/Contents/ServerRoot/usr/sbin:/Library/Apple/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Users/user/.asdf/shims:/opt/homebrew/opt/asdf/libexec/bin:/Users/user/.bin:/Users/user/.antigen/bundles/robbyrussell/oh-my-zsh/lib:/Users/user/.antigen/bundles/robbyrussell/oh-my-zsh/plugins/common-aliases:/Users/user/.antigen/bundles/robbyrussell/oh-my-zsh/plugins/git:/Users/user/.antigen/bundles/robbyrussell/oh-my-zsh/plugins/git-extras:/Users/user/.antigen/bundles/robbyrussell/oh-my-zsh/plugins/git-flow:/Users/user/.antigen/bundles/robbyrussell/oh-my-zsh/plugins/compleat:/Users/user/.antigen/bundles/robbyrussell/oh-my-zsh/plugins/history:/Users/user/.antigen/bundles/robbyrussell/oh-my-zsh/plugins/history-substring-search:/Users/user/.antigen/bundles/zsh-users/zsh-syntax-highlighting:/Users/user/.antigen/bundles/zsh-users/zsh-completions:/Users/user/.antigen/bundles/zsh-users/zsh-autosuggestions:/Users/user/.antigen/bundles/robbyrussell/oh-my-zsh/plugins/fasd:/Users/user/.antigen/bundles/bigbearlabs/dir-scoped-history:/Users/user/.antigen/bundles/robbyrussell/oh-my-zsh/plugins/macos:/Users/user/.antigen/bundles/robbyrussell/oh-my-zsh/plugins/web-search:/Users/user/.antigen/bundles/robbyrussell/oh-my-zsh/plugins/xcode
npm verb lifecycle karabiner.ts.examples@~update: CWD: /Users/user/dev/src/public/karabiner.ts.examples
npm info lifecycle karabiner.ts.examples@~postupdate: karabiner.ts.examples@
npm verb exit [ 0, true ]
npm timing npm Completed in 590ms
npm info ok 
user@host ~/d/s/p/karabiner.ts.examples> npm -v karabiner.ts                                                                                                           main!?
6.14.17
user@host ~/d/s/p/karabiner.ts.examples> npm list                                                                                                                      main!?
karabiner.ts.examples@ /Users/user/dev/src/public/karabiner.ts.examples
├── UNMET PEER DEPENDENCY @types/node@*
├── karabiner.ts@1.4.1 invalid
├── prettier@2.8.8
├─┬ ts-node@10.9.1
│ ├─┬ @cspotcode/source-map-support@0.8.1
│ │ └─┬ @jridgewell/trace-mapping@0.3.9
│ │   ├── @jridgewell/resolve-uri@3.1.1
│ │   └── @jridgewell/sourcemap-codec@1.4.15
│ ├── @tsconfig/node10@1.0.9
│ ├── @tsconfig/node12@1.0.11
│ ├── @tsconfig/node14@1.0.3
│ ├── @tsconfig/node16@1.0.3
│ ├── acorn@8.8.2
│ ├── acorn-walk@8.2.0
│ ├── arg@4.1.3
│ ├── create-require@1.1.1
│ ├── diff@4.0.2
│ ├── make-error@1.3.6
│ ├── v8-compile-cache-lib@3.0.1
│ └── yn@3.1.1
└── typescript@5.0.4

npm ERR! peer dep missing: @types/node@*, required by ts-node@10.9.1
npm ERR! invalid: karabiner.ts@1.4.1 /Users/user/dev/src/public/karabiner.ts.examples/node_modules/karabiner.ts
user@host ~/d/s/p/karabiner.ts.examples>                                                                                                                             1 main!?
```

**Reply by evan-liu** [reply]
Sorry didn't know it would break in old node version, please make the same change as https://github.com/evan-liu/karabiner.ts.examples/commit/89ae6c760696af27c10f18bc1bd762128f994400 (Add `"@types/node": "^20.1.3"` to `"devDependencies"`), then try update again. Sorry about that. (BTW your node version 14 is quite old now, maybe a good idea to upgrade)

**Reply by sohocoke** [reply]
Thanks, that took care of the problem during update.

I'm trying values for `detect_key_down_uninterruptedly` and testing behaviour with 'F+Cmd+J' to select from line end to line start. I'm finding that as soon as I hit Cmd, I get an 'f' char and the layer activation gets messed up.

**Reply by evan-liu** [reply]
I guess layer or simlayer are not right abstractions for your problem. They are both for the pattern of activate a layer then press a key on the layer. You can try the raw mapSimultaneous() with all keys you need. A bit more code to write but should work for your needs.

**Reply by evan-liu** [reply]
Can you set up a repo on GitHub? And I will see if I can make it work and send you a PR.

**Reply by sohocoke** [reply]
forked at https://github.com/sohocoke/karabiner.ts.examples

Really appreciate your help!

**Reply by evan-liu** [reply]
Hi @sohocoke , seems like what you are having to very similar to [what I do](https://github.com/evan-liu/karabiner-config/blob/main/src/layers/arrows.ts). 

```typescript
simlayer('f', 'arrow-layer').manipulators({
  j: toKey('←'),
  i: toKey('↑'),
  k: toKey('↓'),
  l: toKey('→'),

  u: toKey('←', '⌃'),
  o: toKey('→', '⌃'),
})
```

**Comment by evan-liu** [comment]
I do not know a way to allow modifiers in the middle of simlayer, but there are 3 alternative ways, hope one of them works for you:

### Option 1

Use different keys on the layer for adding modifiers. ([I do it this way](https://github.com/evan-liu/karabiner-config/blob/main/src/layers/arrows.ts))

```typescript
simlayer('f', 'arrow-layer').manipulators({
  j: toKey('←'),
  i: toKey('↑'),
  k: toKey('↓'),
  l: toKey('→'),

  u: toKey('←', '⌃'),
  o: toKey('→', '⌃'),
})
```

### Option 2 

Add the modifier to the layer key, then press the modifier + f to activate the 'control-f' layer. 

**Note**: Need to give a different variable name for each modifierLayer, like `'control-f'` in the example. 

```typescript
modifierLayer('⌃', 'f', 'control-f').manipulators({
  u: toKey('←', '⌃'),
  o: toKey('→', '⌃'),
})
```

### Option 3

Press another key close to 'f' key as the modifier key. Then press the key together with 'f' to activate the layer. 

```typescript 
duoLayer('f', 'd').manipulators({
  u: toKey('←', '⌃'),
  o: toKey('→', '⌃'),
})
```

**Comment by evan-liu** [comment]
Hi @sohocoke, in case you still need it, there seems to be a new option now with https://github.com/evan-liu/karabiner.ts/issues/89 in `1.20.0`.
